How many late days are you using for this assignment? 0 Late Days

In [911]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

# Part 1: Load the dataset

In [912]:
# Load the given datasets
ckd_categorical = pd.read_csv('chronic_kidney_disease_categorical.csv')
ckd_numerical = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
print(ckd_categorical)
print(ckd_numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
The data is medical data about specific patients, specifically used to predict whether a patient has chronic kidney disease or not based on other data. The only required data is a binary (boolean) showing whether or not the patient has chronic kidney disease. All other features of the data are optional for each patient, and they include binary, categorical, integer, and continuous values. Every numerical value (both integers and continuous) have a label unit attached (always some volume, or quantity per volume, except age). 

3 Examples of categorical columns would be appetite (whether the patient has a normal appetite or not), hypertension (medical condition, which my Dad has), or whether bacteria is present or not.

3 Examples of numerical columns would be Blood Urea (mgs/dl), Sodium (mEq/L), and Potassium(mEq/L).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [913]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print("There are " + str(len(ckd_numerical)) + " rows in the numerical dataset") 
numerical_duplicate_count = ckd_numerical['unique_id'].duplicated(keep=False).sum()
print("There are " + str(numerical_duplicate_count) + " duplicate rows in the numerical dataset")
ckd_numerical_duplicates = ckd_numerical[ckd_numerical.duplicated(keep=False)]
print("Double check number of duplicates: " + str(len(ckd_numerical_duplicates)))

# Drop these duplicate rows
ckd_numerical_no_duplicates = ckd_numerical.drop_duplicates(subset = 'unique_id', keep = 'first')
print("With duplicates dropped, there are " + str(len(ckd_numerical_no_duplicates)) + " rows left in the numerical dataset")

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print("There are " + str(len(ckd_categorical)) + " rows in the categorical dataset") 
categorical_duplicate_count = ckd_categorical['unique_id'].duplicated(keep=False).sum()
print("There are " + str(categorical_duplicate_count) + " duplicate rows in the categorical dataset")

ckd_categorical_no_duplicates = ckd_categorical.drop_duplicates(subset = 'unique_id', keep = 'first')
print("With duplicates dropped, there are " + str(len(ckd_categorical_no_duplicates)) + " rows left in the categorical dataset")



There are 420 rows in the numerical dataset
There are 59 duplicate rows in the numerical dataset
Double check number of duplicates: 59
With duplicates dropped, there are 390 rows left in the numerical dataset
There are 430 rows in the categorical dataset
There are 59 duplicate rows in the categorical dataset
With duplicates dropped, there are 400 rows left in the categorical dataset


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
All the unique ids are not present in both datasets, we know this because the size of the categorical dataset with all duplicates removed is greater than the size of the numerical dataset with all duplicates removed. The rows that are missing from one of the datasets will have the columns from that dataset all empty (NaN).

In [914]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_dataset = pd.merge(ckd_numerical_no_duplicates, ckd_categorical_no_duplicates, on='unique_id', how='outer')
#Print the combined dataset
print(merged_dataset)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [915]:
# Calculate the percentage of rows that contain atleast one missing value
missing_value_count = merged_dataset.isna().any(axis=1).sum()
missing_value_percentage = missing_value_count / len(merged_dataset) * 100
# Print %
print(missing_value_percentage)
# Drop these rows from the dataset
complete_only = merged_dataset.dropna()
# Print the Dataset
print(complete_only)

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc        

## Part 3.4: Sort the dataset according to the Labels

In [916]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_by_target = complete_only.sort_values(by = 'Target').reset_index().drop('index', axis = 1)

# Print the dataset
print(sorted_by_target)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [917]:
# Write code here, note: should add 11 extra columns
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dummied_data = pd.get_dummies(sorted_by_target, columns = categorical_columns)
# Print the dataset
print(dummied_data)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: In this case it'd be equivalent in downtsream modeling, since every categorical data has only true/false values. So for every category dummy column, there's another column next to it with the opposite value in every row. If the categorical variable could take more than 2 values, this would no longer work. We could instead however, map each category to an integer value and then encode that integer value in a single dummy column (rather than encoding as a binary).

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [918]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def remove_outliers(dataframe, columns):
    df_filtered = dataframe.copy()
    for column in columns:
        mean = df_filtered[column].mean()
        std_dev = df_filtered[column].std()
        z_scores = (df_filtered[column] - mean) / std_dev
        outliers = np.abs(z_scores) > 3
        df_filtered = df_filtered[~outliers]
    return df_filtered
no_outliers = remove_outliers(dummied_data, numerical_columns)
# Print the dataset
print(no_outliers)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [919]:
# Normalize the all Numerical Attributes in the dataset.
#Normalizes values in column to 0-100 range
def min_max_normalize(column):
    return (column - column.min()) / (column.max() - column.min()) * 100
for col in numerical_columns:
    no_outliers[f'{col}'] = min_max_normalize(no_outliers[col])
# Print the dataset
print(no_outliers)


     unique_id        age     bp        bgr         bu         sc         sod  \
2       343710  74.324324   50.0  44.206009  90.196078  43.209877   50.000000   
5       397388  77.027027  100.0  90.128755  16.339869  34.567901   76.666667   
7       438182  90.540541   50.0  78.540773  86.274510  51.851852   60.000000   
8       474407  20.270270   75.0  15.879828  19.607843  16.049383   16.666667   
9       475200  78.378378  100.0  39.914163  28.758170  83.950617   66.666667   
..         ...        ...    ...        ...        ...        ...         ...   
148     486397  18.918919   25.0  22.746781  22.222222   7.407407   50.000000   
149     203466  21.621622    0.0  11.587983   5.228758   9.876543   60.000000   
150     482105  51.351351    0.0  10.729614  23.529412   1.234568   60.000000   
151     228849  70.270270   50.0  12.875536  26.143791   9.876543   66.666667   
152     996076  68.918919    0.0  15.021459  25.490196   9.876543  100.000000   

            pot       hemo 

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column has absolutely nothing to do with a person's health, just a random identifier number. Should not be fed as input for modeling.

In [920]:
#Remove that column
clean_data = no_outliers.drop('unique_id', axis = 1).reset_index().drop('index', axis = 1)
# Print the dataset
print(clean_data)

           age     bp        bgr         bu         sc         sod  \
0    74.324324   50.0  44.206009  90.196078  43.209877   50.000000   
1    77.027027  100.0  90.128755  16.339869  34.567901   76.666667   
2    90.540541   50.0  78.540773  86.274510  51.851852   60.000000   
3    20.270270   75.0  15.879828  19.607843  16.049383   16.666667   
4    78.378378  100.0  39.914163  28.758170  83.950617   66.666667   
..         ...    ...        ...        ...        ...         ...   
128  18.918919   25.0  22.746781  22.222222   7.407407   50.000000   
129  21.621622    0.0  11.587983   5.228758   9.876543   60.000000   
130  51.351351    0.0  10.729614  23.529412   1.234568   60.000000   
131  70.270270   50.0  12.875536  26.143791   9.876543   66.666667   
132  68.918919    0.0  15.021459  25.490196   9.876543  100.000000   

            pot       hemo        pcv       wbcc  ...  cad_no  cad_yes  \
0     79.310345   0.000000   3.225806  39.516129  ...   False     True   
1     20.68

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [921]:
# Export the dataframe to a new csv file
clean_data.to_csv('cleaned_data.csv', index = False)
# Export the dataframe to a new json file
clean_data.to_json('cleaned_data.json', index = False)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [922]:
for i in range(15):
    print(clean_data.iloc[i])
    print("")

age               74.324324
bp                     50.0
bgr               44.206009
bu                90.196078
sc                43.209877
sod                    50.0
pot               79.310345
hemo                    0.0
pcv                3.225806
wbcc              39.516129
rbcc               5.714286
al                      2.0
su                      0.0
rbc_abnormal           True
rbc_normal            False
pc_abnormal            True
pc_normal             False
pcc_notpresent         True
pcc_present           False
ba_notpresent          True
ba_present            False
htn_no                False
htn_yes                True
dm_no                 False
dm_yes                 True
cad_no                False
cad_yes                True
appet_good            False
appet_poor             True
pe_no                 False
pe_yes                 True
ane_no                False
ane_yes                True
Target_ckd             True
Target_notckd         False
Name: 0, dtype: obje

### Paste the markdown table here

| age        | bp   | bgr       | bu        | sc       | sod      | pot       | hemo      | pcv      | wbcc      | rbcc      | al  | su  | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------------|------|-----------|-----------|----------|----------|-----------|-----------|----------|-----------|-----------|-----|-----|---------------|------------|-------------|-----------|----------------|-------------|---------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|-------------|----------------|
| 74.324324  | 50.0 | 44.206009 | 90.196078 | 43.209877| 50.0     | 79.310345 | 0.0       | 3.225806 | 39.516129 | 5.714286  | 2.0 | 0.0 | True          | False      | True        | False     | True           | False       | True          | False      | False  | True    | False | True   | False  | True    | False      | True       | False | True   | False  | True    | True        | False          |
| 77.027027  | 100.0| 90.128755 | 16.339869 | 34.567901| 76.666667| 20.689655 | 52.475248 | 54.83871 | 44.354839 | 34.285714 | 2.0 | 2.0 | False         | True       | False       | True      | True           | False       | False         | True       | False  | True    | True  | False  | False  | True    | True       | False      | True   | False  | True   | False   | True        | False          |
| 90.540541  | 50.0 | 78.540773 | 86.27451  | 51.851852| 60.0     | 100.0     | 27.722772 | 32.258065| 23.387097 | 37.142857 | 2.0 | 0.0 | True          | False      | True        | False     | True           | False       | True          | False      | False  | True    | False | True   | False  | True    | True       | False      | True   | False  | True   | False   | True        | False          |
| 20.27027   | 75.0 | 15.879828 | 19.607843 | 16.049383| 16.666667| 20.689655 | 5.940594  | 0.0      | 65.322581 | 25.714286 | 4.0 | 0.0 | False         | True       | True        | False     | False          | True        | False         | True       | True   | False   | True  | False  | True   | False   | True       | False      | True   | False  | False  | True    | True        | False          |
| 78.378378  | 100.0| 39.914163 | 28.75817  | 83.950617| 66.666667| 58.62069  | 1.980198  | 9.677419 | 25.806452 | 11.428571 | 4.0 | 2.0 | True          | False      | True        | False     | True           | False       | False         | True       | False  | True    | False | True   | True   | False   | True       | False      | False  | True   | True   | False   | True        | False          |
| 72.972973  | 0.0  | 93.562232 | 16.993464 | 16.049383| 33.333333| 3.448276  | 1.980198  | 6.451613 | 87.903226 | 0.0       | 3.0 | 1.0 | False         | True       | True        | False     | False          | True        | True          | False      | False  | True    | True  | False  | True   | False   | False      | True       | True   | False  | False  | True    | True        | False          |
| 67.567568  | 75.0 | 25.321888 | 63.398693 | 77.777778| 36.666667| 65.517241 | 13.861386 | 19.354839| 16.935484 | 11.428571 | 2.0 | 0.0 | True          | False      | True        | False     | True           | False       | True          | False      | False  | True    | True  | False  | True   | False   | True       | False      | True   | False  | True   | False   | True        | False          |
| 85.135135  | 25.0 | 61.802575 | 56.20915  | 72.839506| 0.0      | 34.482759 | 16.831683 | 16.129032| 58.064516 | 8.571429  | 4.0 | 3.0 | False         | True       | True        | False     | False          | True        | False         | True       | False  | True    | False | True   | False  | True    | True       | False      | False  | True   | False  | True    | True        | False          |
| 54.054054  | 0.0  | 39.914163 | 53.594771 | 35.802469| 70.0     | 37.931034 | 20.792079 | 16.129032| 83.064516 | 5.714286  | 1.0 | 0.0 | False         | True       | False       | True      | True           | False       | True          | False      | False  | True    | False | True   | True   | False   | True       | False      | True   | False  | True   | False   | True        | False          |
| 75.675676  | 25.0 | 22.317597 | 20.915033 | 16.049383| 53.333333| 62.068966 | 48.514851 | 51.612903| 29.032258 | 25.714286 | 3.0 | 0.0 | False         | True       | True        | False     | True           | False       | True          | False      | False  | True    | False | True   | True   | False   | True       | False      | True   | False  | True   | False   | True        | False          |
| 66.216216  | 50.0 | 61.802575 | 41.176471 | 43.209877| 56.666667| 68.965517 | 31.683168 | 35.483871| 25.0      | 20.0      | 3.0 | 1.0 | False         | True       | True        | False     | False          | True        | False         | True       | False  | True    | False | True   | True   | False   | True       | False      | True   | False  | True   | False   | True        | False          |
| 78.378378  | 0.0  | 72.532189 | 31.372549 | 48.148148| 56.666667| 86.206897 | 17.821782 | 19.354839| 25.806452 | 11.428571 | 4.0 | 1.0 | True          | False      | True        | False     | True           | False       | False         | True       | False  | True    | False | True   | True   | False   | True       | False      | True   | False  | True   | False   | True        | False          |
| 87.837838  | 0.0  | 20.600858 | 75.163399 | 60.493827| 53.333333| 68.965517 | 36.633663 | 38.709677| 87.903226 | 37.142857 | 4.0 | 0.0 | False         | True       | False       | True      | True           | False       | True          | False      | False  | True    | False | True   | True   | False   | True       | False      | True   | False  | True   | False   | True        | False          |
| 87.837838  | 25.0 | 63.948498 | 47.058824 | 39.506173| 43.333333| 51.724138 | 26.732673 | 32.258065| 10.483871 | 17.142857 | 3.0 | 0.0 | False         | True       | True        | False     | False          | True        | False         | True       | False  | True    | False | True   | True   | False   | True       | False      | True   | False  | True   | False   | True        | False          |
| 71.621622  | 50.0 | 100.0     | 16.339869 | 11.111111| 6.666667 | 20.689655 | 26.732673 | 38.709677| 53.225806 | 37.142857 | 1.0 | 0.0 | True          | False      | False       | True      | True           | False       | True          | False      | True   | False   | False | True   | True   | False   | False      | True       | True   | False  | True   | False   | True        | False          |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Batch the data into smaller sets, feeding them one set at a time. Performing random checks on values would be useful, as well as checking that the overall number of rows is correct.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [923]:
# Code to rename all the columns in the dataset
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rc': 'red blood cell count',
    'class': 'class'
}

# Duplicating and adding suffixes to specific keys
additional_mappings = {
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'Target': 'target'
}

# Creating the new mappings with the suffixes
for key, value in additional_mappings.items():
    column_mapping[f"{key}_no"] = f"Does not have {value}"
    column_mapping[f"{key}_yes"] = f"Has {value}"
    column_mapping[f"{key}_poor"] = f"{value} is poor"
    column_mapping[f"{key}_good"] = f"{value} is good"
    column_mapping[f"{key}_ckd"] = f"{value} tests positive for chronic kidney disease"
    column_mapping[f"{key}_notckd"] = f"{value} tests negative for chronic kidney disease"
    column_mapping[f"{key}_abnormal"] = f"{value} is abnormal"
    column_mapping[f"{key}_normal"] = f"{value} is normal"
    column_mapping[f"{key}_present"] = f"{value} is present"
    column_mapping[f"{key}_notpresent"] = f"{value} is not present"

# Rename the columns using the mapping
clean_data.rename(columns=column_mapping, inplace=True)
print(clean_data)

           age  blood pressure  blood glucose random  blood urea  \
0    74.324324            50.0             44.206009   90.196078   
1    77.027027           100.0             90.128755   16.339869   
2    90.540541            50.0             78.540773   86.274510   
3    20.270270            75.0             15.879828   19.607843   
4    78.378378           100.0             39.914163   28.758170   
..         ...             ...                   ...         ...   
128  18.918919            25.0             22.746781   22.222222   
129  21.621622             0.0             11.587983    5.228758   
130  51.351351             0.0             10.729614   23.529412   
131  70.270270            50.0             12.875536   26.143791   
132  68.918919             0.0             15.021459   25.490196   

     serum creatinine      sodium   potassium  hemoglobin  packed cell volume  \
0           43.209877   50.000000   79.310345    0.000000            3.225806   
1           34.567901

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [924]:
# Converted SQL to Pandas code
# Sample data: You would typically load this from a CSV or a database.
data = {
    'Target': ['ckd', 'notckd', 'ckd', 'ckd', 'notckd', 'notckd', 'ckd', 'notckd'],
    # Include other columns if necessary
}

# Create the DataFrame from the sample data
df = pd.DataFrame(data)

# Perform the grouping and counting
result = df.groupby('Target').size().reset_index(name='count')

# Display the resulting DataFrame
print(result)

   Target  count
0     ckd      4
1  notckd      4


*Describe what the above code does here*

Counts the amount that unique values appear for a certain column.